File: Polynomial Regression/polynomial_regression.py
# polynomial regression

# Importing the libraries
import numpy as np  # math tools
import matplotlib.pyplot as plt  # plotting
import pandas as pd  # import and manage data/datasets

# Importing the dataset
dataset = pd.read_csv('data/Position_Salaries.csv')
# even though there is one value for X, by calling a second column you ensure you have a matrix, not a vector.
X = dataset.iloc[:, 1:2].values # independent variable - level
y = dataset.iloc[:, 2].values # last column. Dependent variable

# define 4 variables - X_train, X_test, y_train, y_test - at the same time
# in this case the data set is too small, so we want to train on everything.
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Fitting linear regression to the dataset 
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting Polynomial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4) # changing degrees will change the fit of the line. Mess around with it.
X_poly = poly_reg.fit_transform(X) # X_poly column 1 = intercept
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly, y)

# Visualising Linear Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visualising the Polynomial Regression results
X_grid = np.arange(min(X), max(X), 0.1) 
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_reg2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# predicting a new result with Linear Regression
lin_reg.predict(np.array(6.5).reshape(-1, 1))

# predicitng a new result with polynomial regresson
lin_reg2.predict(poly_reg.fit_transform(np.array(6.5).reshape(-1, 1)))

File: Polynomial Regression/polynomial_regression.R
# polynomial regression

# Importing the data set
dataset <- read.csv("data/position_salaries.csv")

# remove the first column
dataset <- dataset[,2:3]

# splitting data in train and test sets - not needed for this exercise as datas et is too small
# library(caTools)
# set.seed(123)
# split <- caTools::sample.split(dataset$Purchased, SplitRatio = 0.8) # returns 80% TRUE and 20% FALSE
# training_set <- subset(dataset, split == TRUE) # Take TRUE values
# test_set <- subset(dataset, split == FALSE) # Take false values
# 
# # feature scaling - need for this either
# training_set[,2:3] <- scale(training_set[,2:3])
# test_set[,2:3] <- scale(test_set[,2:3])

# Fitting the linear regression to the data set
lin_reg <- lm(formula = Salary ~ ., 
              data = dataset)

summary(lin_reg)

# Fitting the Polynomial Regression to the data set

# polynomial regression requires additional independent variables that are 
# polynomial terms of the 1st independent variable. More 'degree' variables help fit the model better
dataset$Level2 <- dataset$Level^2 
dataset$Level3 <- dataset$Level^3 
dataset$Level4 <- dataset$Level^4

poly_reg <- lm(formula = Salary ~ .,
               data = dataset)

summary(poly_reg)

# Visualizing the Linear Regression Results
library(ggplot2)

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), # real observation points
                 colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), # predicted observation points
            colour = 'blue') +
  ggtitle("Truth or Bluff (Linear Regresson)") + 
  xlab('Level') +
  ylab('Salary')
  
# Visualizing the Polynomial Regression Results
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), # real observation points
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)), # predicted observation points
            colour = 'blue') +
  ggtitle("Truth or Bluff (Polynomial Regresson)") + 
  xlab('Level') +
  ylab('Salary')

# Predicting a new result with linear regression
# create a new data set to examine a single data point

y_pred <- predict(lin_reg, data.frame(Level = 6.5))

# Predicting a new result with polynomial regression
# create a new data set to examine a single data point

y_pred <- predict(poly_reg, data.frame(Level = 6.5,
                                       Level2 = 6.5^2,
                                       Level3 = 6.5^3,
                                       Level4 = 6.5^4)) # you must input as many values as there are degrees in the poly_reg model


